File: BattleboatIOS/BattleboatIOS/AI.swift
//
//  AI.swift
//  BattleboatIOS
//
//  Created by Battleboat on 2024/01/01.
//  Copyright © 2024 Battleboat. All rights reserved.
//

import Foundation

class AI {
    private let grid: Grid
    private let fleet: Fleet
    private var probabilityGrid: [[Double]]
    private var huntMode: Bool = false
    private var huntTargets: [(x: Int, y: Int)] = []
    private var lastHit: (x: Int, y: Int)?
    private var huntDirection: (dx: Int, dy: Int)?
    
    init(grid: Grid, fleet: Fleet) {
        self.grid = grid
        self.fleet = fleet
        self.probabilityGrid = Array(repeating: Array(repeating: 0.0, count: GameConstants.gridSize), count: GameConstants.gridSize)
        calculateInitialProbabilities()
    }
    
    // MARK: - AI Decision Making
    
    func makeMove() -> (x: Int, y: Int) {
        updateProbabilities()
        
        if huntMode && !huntTargets.isEmpty {
            // In hunt mode, target adjacent cells to known hits
            return selectHuntTarget()
        } else {
            // In search mode, use probability-based targeting
            return selectBestProbabilityTarget()
        }
    }
    
    func processResult(x: Int, y: Int, result: GameConstants.CellType, sunkShip: Ship?) {
        switch result {
        case .hit:
            handleHit(x: x, y: y)
        case .miss:
            handleMiss(x: x, y: y)
        case .sunk:
            handleSunk(x: x, y: y, ship: sunkShip)
        default:
            break
        }
        
        updateProbabilities()
    }
    
    // MARK: - Hunt Mode Logic
    
    private func handleHit(x: Int, y: Int) {
        lastHit = (x: x, y: y)
        huntMode = true
        
        // Add adjacent cells to hunt targets
        let adjacent = grid.getAdjacentCoordinates(x: x, y: y, includeDiagonals: false)
        for coord in adjacent {
            let cellType = grid.getCellType(x: coord.x, y: coord.y)
            if (cellType == .empty || cellType == .ship) && !huntTargets.contains(where: { $0.x == coord.x && $0.y == coord.y }) {
                huntTargets.append(coord)
            }
        }
        
        // If we have a previous hit in a line, prioritize that direction
        updateHuntDirection(newHit: (x: x, y: y))
    }
    
    private func handleMiss(x: Int, y: Int) {
        // Remove from hunt targets if it was there
        huntTargets.removeAll { $0.x == x && $0.y == y }
        
        // If we're in hunt mode and this was following a direction, reverse direction
        if let direction = huntDirection, let lastHit = lastHit {
            let oppositeDirection = (dx: -direction.dx, dy: -direction.dy)
            let oppositeTarget = (x: lastHit.x + oppositeDirection.dx, y: lastHit.y + oppositeDirection.dy)
            
            if grid.isValidCoordinate(x: oppositeTarget.x, y: oppositeTarget.y) {
                let cellType = grid.getCellType(x: oppositeTarget.x, y: oppositeTarget.y)
                if cellType == .empty || cellType == .ship {
                    huntTargets.insert(oppositeTarget, at: 0) // Priority target
                }
            }
        }
    }
    
    private func handleSunk(x: Int, y: Int, ship: Ship?) {
        // Clear hunt mode
        huntMode = false
        huntTargets.removeAll()
        lastHit = nil
        huntDirection = nil
        
        // Remove cells around the sunk ship from future consideration
        if let ship = ship {
            for cell in ship.getAllShipCells() {
                let adjacent = grid.getAdjacentCoordinates(x: cell.x, y: cell.y, includeDiagonals: true)
                for coord in adjacent {
                    let cellType = grid.getCellType(x: coord.x, y: coord.y)
                    if cellType == .empty || cellType == .ship {
                        // These cells can't contain ships (ships don't touch)
                        // We'll reduce their probability significantly
                    }
                }
            }
        }
    }
    
    private func updateHuntDirection(newHit: (x: Int, y: Int)) {
        guard let lastHit = lastHit else { return }
        
        let dx = newHit.x - lastHit.x
        let dy = newHit.y - lastHit.y
        
        if abs(dx) == 1 && dy == 0 {
            huntDirection = (dx: dx, dy: 0)
        } else if abs(dy) == 1 && dx == 0 {
            huntDirection = (dx: 0, dy: dy)
        }
    }
    
    private func selectHuntTarget() -> (x: Int, y: Int) {
        // Prioritize targets in the hunt direction if we have one
        if let direction = huntDirection, let lastHit = lastHit {
            let priorityTarget = (x: lastHit.x + direction.dx, y: lastHit.y + direction.dy)
            
            if grid.isValidCoordinate(x: priorityTarget.x, y: priorityTarget.y) {
                let cellType = grid.getCellType(x: priorityTarget.x, y: priorityTarget.y)
                if cellType == .empty || cellType == .ship {
                    huntTargets.removeAll { $0.x == priorityTarget.x && $0.y == priorityTarget.y }
                    return priorityTarget
                }
            }
        }
        
        // Otherwise, select the first available hunt target
        if let target = huntTargets.first {
            huntTargets.removeFirst()
            return target
        }
        
        // Fall back to probability-based selection
        huntMode = false
        return selectBestProbabilityTarget()
    }
    
    // MARK: - Probability Calculation
    
    private func calculateInitialProbabilities() {
        // Reset probability grid
        for x in 0..<GameConstants.gridSize {
            for y in 0..<GameConstants.gridSize {
                probabilityGrid[x][y] = 0.0
            }
        }
        
        // Calculate probabilities for each ship type
        for shipType in GameConstants.ShipType.allCases {
            calculateProbabilitiesForShip(shipType: shipType)
        }
    }
    
    private func updateProbabilities() {
        calculateInitialProbabilities()
        applyHeuristicModifiers()
    }
    
    private func calculateProbabilitiesForShip(shipType: GameConstants.ShipType) {
        let shipSize = shipType.size
        
        // Try placing the ship in every possible position and orientation
        for x in 0..<GameConstants.gridSize {
            for y in 0..<GameConstants.gridSize {
                for direction in [GameConstants.ShipDirection.horizontal, GameConstants.ShipDirection.vertical] {
                    if canPlaceVirtualShip(x: x, y: y, size: shipSize, direction: direction) {
                        // Add probability weight to all cells this ship would occupy
                        let coords = calculateShipCoordinates(x: x, y: y, size: shipSize, direction: direction)
                        for coord in coords {
                            probabilityGrid[coord.x][coord.y] += 1.0
                        }
                    }
                }
            }
        }
    }
    
    private func canPlaceVirtualShip(x: Int, y: Int, size: Int, direction: GameConstants.ShipDirection) -> Bool {
        let coords = calculateShipCoordinates(x: x, y: y, size: size, direction: direction)
        
        var hitCount = 0
        
        for coord in coords {
            // Check bounds
            if !grid.isValidCoordinate(x: coord.x, y: coord.y) {
                return false
            }
            
            let cellType = grid.getCellType(x: coord.x, y: coord.y)
            
            // Can't place on miss or sunk cells
            if cellType == .miss || cellType == .sunk {
                return false
            }
            
            // Count hits - ship must account for all hits in its path
            if cellType == .hit {
                hitCount += 1
            }
        }
        
        // For hunt mode, prioritize ships that contain known hits
        if huntMode && hitCount == 0 {
            return false
        }
        
        return true
    }
    
    private func calculateShipCoordinates(x: Int, y: Int, size: Int, direction: GameConstants.ShipDirection) -> [(x: Int, y: Int)] {
        var coords: [(x: Int, y: Int)] = []
        
        for i in 0..<size {
            if direction == .horizontal {
                coords.append((x: x + i, y: y))
            } else {
                coords.append((x: x, y: y + i))
            }
        }
        
        return coords
    }
    
    private func applyHeuristicModifiers() {
        for x in 0..<GameConstants.gridSize {
            for y in 0..<GameConstants.gridSize {
                // Don't target cells we've already shot
                let cellType = grid.getCellType(x: x, y: y)
                if cellType == .miss || cellType == .hit || cellType == .sunk {
                    probabilityGrid[x][y] = 0.0
                    continue
                }
                
                // Boost probability for cells adjacent to hits
                let adjacent = grid.getAdjacentCoordinates(x: x, y: y, includeDiagonals: false)
                for coord in adjacent {
                    if grid.isHit(x: coord.x, y: coord.y) {
                        probabilityGrid[x][y] *= 2.0
                    }
                }
                
                // Reduce probability near sunk ships (ships don't touch)
                let adjacentWithDiagonals = grid.getAdjacentCoordinates(x: x, y: y, includeDiagonals: true)
                for coord in adjacentWithDiagonals {
                    if grid.isSunk(x: coord.x, y: coord.y) {
                        probabilityGrid[x][y] *= 0.1
                    }
                }
                
                // Apply checkerboard pattern for parity (ships must be hit)
                // This gives slight preference to cells that follow ship placement patterns
                if (x + y) % 2 == 0 {
                    probabilityGrid[x][y] *= 1.1
                }
            }
        }
    }
    
    private func selectBestProbabilityTarget() -> (x: Int, y: Int) {
        var bestX = 0
        var bestY = 0
        var maxProbability = -1.0
        
        for x in 0..<GameConstants.gridSize {
            for y in 0..<GameConstants.gridSize {
                let cellType = grid.getCellType(x: x, y: y)
                if (cellType == .empty || cellType == .ship) && probabilityGrid[x][y] > maxProbability {
                    maxProbability = probabilityGrid[x][y]
                    bestX = x
                    bestY = y
                }
            }
        }
        
        return (x: bestX, y: bestY)
    }
    
    // MARK: - Debugging
    
    func printProbabilityGrid() {
        print("AI Probability Grid:")
        for y in 0..<GameConstants.gridSize {
            var row = ""
            for x in 0..<GameConstants.gridSize {
                let prob = probabilityGrid[x][y]
                let cellType = grid.getCellType(x: x, y: y)
                if cellType == .empty || cellType == .ship {
                    row += String(format: "%3.0f ", prob)
                } else {
                    row += "  X "
                }
            }
            print(row)
        }
        print("")
    }
    
    func getProbabilityAt(x: Int, y: Int) -> Double {
        guard grid.isValidCoordinate(x: x, y: y) else { return 0.0 }
        return probabilityGrid[x][y]
    }
} 

File: BattleboatAndroid/app/src/main/java/com/battleboat/AI.kt
package com.battleboat

import kotlin.math.max
import kotlin.math.min
import kotlin.random.Random

/**
 * Advanced AI opponent for Battleboat game
 * Implements probability-based targeting with hunt mode
 */
class AI {
    private val probabilityGrid = Array(GameConstants.GRID_SIZE) { 
        Array(GameConstants.GRID_SIZE) { 0 } 
    }
    
    private val targetStack = mutableListOf<Pair<Int, Int>>()
    private var huntMode = false
    private var lastHit: Pair<Int, Int>? = null
    private var huntDirection: Pair<Int, Int>? = null
    private var huntOrigin: Pair<Int, Int>? = null
    
    private var difficulty = AIDifficulty.MEDIUM
    private var shotCount = 0
    private var hitCount = 0
    
    /**
     * Set AI difficulty level
     */
    fun setDifficulty(difficulty: AIDifficulty) {
        this.difficulty = difficulty
    }
    
    /**
     * Reset AI state for new game
     */
    fun reset() {
        clearProbabilityGrid()
        targetStack.clear()
        huntMode = false
        lastHit = null
        huntDirection = null
        huntOrigin = null
        shotCount = 0
        hitCount = 0
    }
    
    /**
     * Get the next target position for the AI
     */
    fun getNextTarget(playerGrid: Grid, aiFleet: Fleet): Pair<Int, Int>? {
        shotCount++
        
        // Update probability grid based on current state
        updateProbabilityGrid(playerGrid, aiFleet)
        
        // If in hunt mode, continue hunting
        if (huntMode && targetStack.isNotEmpty()) {
            return getNextHuntTarget()
        }
        
        // Get best target based on probability
        return getBestTarget(playerGrid)
    }
    
    /**
     * Process the result of a shot
     */
    fun processShotResult(target: Pair<Int, Int>, result: ShotResult, sunkShip: Ship?) {
        val (row, col) = target
        
        when (result) {
            ShotResult.HIT -> {
                hitCount++
                onHit(row, col)
                
                // If ship was sunk, exit hunt mode
                if (sunkShip != null) {
                    onShipSunk(sunkShip)
                }
            }
            ShotResult.MISS -> {
                onMiss(row, col)
            }
            else -> {}
        }
    }
    
    /**
     * Handle successful hit
     */
    private fun onHit(row: Int, col: Int) {
        if (!huntMode) {
            // Enter hunt mode
            huntMode = true
            huntOrigin = row to col
            huntDirection = null
            addAdjacentTargets(row, col)
        } else {
            // Continue hunt mode
            lastHit?.let { (lastRow, lastCol) ->
                if (huntDirection == null) {
                    // Determine direction
                    val deltaRow = row - lastRow
                    val deltaCol = col - lastCol
                    huntDirection = deltaRow to deltaCol
                    
                    // Add targets in this direction
                    addDirectionalTargets(row, col, huntDirection!!)
                    
                    // Also add targets in opposite direction from origin
                    huntOrigin?.let { (originRow, originCol) ->
                        val oppositeDelta = -deltaRow to -deltaCol
                        addDirectionalTargets(originRow, originCol, oppositeDelta)
                    }
                } else {
                    // Continue in same direction
                    addDirectionalTargets(row, col, huntDirection!!)
                }
            }
        }
        
        lastHit = row to col
    }
    
    /**
     * Handle missed shot
     */
    private fun onMiss(row: Int, col: Int) {
        // If we were hunting in a direction, try the opposite direction
        if (huntMode && huntDirection != null) {
            huntOrigin?.let { (originRow, originCol) ->
                val (deltaRow, deltaCol) = huntDirection!!
                val oppositeDelta = -deltaRow to -deltaCol
                addDirectionalTargets(originRow, originCol, oppositeDelta)
            }
            huntDirection = null
        }
    }
    
    /**
     * Handle ship sunk
     */
    private fun onShipSunk(ship: Ship) {
        huntMode = false
        targetStack.clear()
        lastHit = null
        huntDirection = null
        huntOrigin = null
        
        // Mark area around sunk ship as low priority
        markSunkShipArea(ship)
    }
    
    /**
     * Add adjacent targets to hunt stack
     */
    private fun addAdjacentTargets(row: Int, col: Int) {
        val adjacentPositions = GameConstants.getAdjacentPositions(row, col)
        for (pos in adjacentPositions) {
            if (!targetStack.contains(pos)) {
                targetStack.add(0, pos) // Add to front for priority
            }
        }
    }
    
    /**
     * Add directional targets to hunt stack
     */
    private fun addDirectionalTargets(row: Int, col: Int, direction: Pair<Int, Int>) {
        val (deltaRow, deltaCol) = direction
        var currentRow = row + deltaRow
        var currentCol = col + deltaCol
        
        // Add up to 4 targets in this direction
        for (i in 0 until 4) {
            if (GameConstants.isValidPosition(currentRow, currentCol)) {
                val target = currentRow to currentCol
                if (!targetStack.contains(target)) {
                    targetStack.add(0, target)
                }
                currentRow += deltaRow
                currentCol += deltaCol
            } else {
                break
            }
        }
    }
    
    /**
     * Get next target from hunt stack
     */
    private fun getNextHuntTarget(): Pair<Int, Int>? {
        while (targetStack.isNotEmpty()) {
            val target = targetStack.removeAt(0)
            // Remove targets that have already been shot
            // This will be checked by the game logic
            return target
        }
        huntMode = false
        return null
    }
    
    /**
     * Get best target based on probability
     */
    private fun getBestTarget(playerGrid: Grid): Pair<Int, Int>? {
        val availableTargets = playerGrid.getAvailableTargets()
        if (availableTargets.isEmpty()) return null
        
        // Apply difficulty-based randomization
        val candidates = when (difficulty) {
            AIDifficulty.EASY -> {
                // Random selection with slight bias toward high probability
                val shuffled = availableTargets.shuffled()
                shuffled.take(min(shuffled.size, 10))
            }
            AIDifficulty.MEDIUM -> {
                // Select from top 30% of targets
                val sorted = availableTargets.sortedByDescending { (row, col) ->
                    probabilityGrid[row][col]
                }
                val topCount = max(1, (sorted.size * 0.3).toInt())
                sorted.take(topCount)
            }
            AIDifficulty.HARD -> {
                // Always select highest probability targets
                val maxProbability = availableTargets.maxOfOrNull { (row, col) ->
                    probabilityGrid[row][col]
                } ?: 0
                
                availableTargets.filter { (row, col) ->
                    probabilityGrid[row][col] == maxProbability
                }
            }
        }
        
        return candidates.randomOrNull()
    }
    
    /**
     * Update probability grid based on current game state
     */
    private fun updateProbabilityGrid(playerGrid: Grid, aiFleet: Fleet) {
        clearProbabilityGrid()
        
        // Get remaining ship sizes
        val remainingShipSizes = getRemainingShipSizes(aiFleet)
        
        // Calculate probability for each cell
        for (row in 0 until GameConstants.GRID_SIZE) {
            for (col in 0 until GameConstants.GRID_SIZE) {
                if (!playerGrid.hasBeenShot(row, col)) {
                    probabilityGrid[row][col] = calculateCellProbability(
                        row, col, playerGrid, remainingShipSizes
                    )
                }
            }
        }
        
        // Apply hunt mode bonuses
        if (huntMode) {
            applyHuntModeBonuses(playerGrid)
        }
        
        // Apply strategic bonuses
        applyStrategicBonuses(playerGrid)
    }
    
    /**
     * Calculate probability for a specific cell
     */
    private fun calculateCellProbability(
        row: Int, 
        col: Int, 
        playerGrid: Grid, 
        shipSizes: List<Int>
    ): Int {
        var probability = 0
        
        for (shipSize in shipSizes) {
            // Check horizontal placement
            if (canPlaceShipHorizontally(row, col, shipSize, playerGrid)) {
                probability += shipSize
            }
            
            // Check vertical placement
            if (canPlaceShipVertically(row, col, shipSize, playerGrid)) {
                probability += shipSize
            }
        }
        
        return probability
    }
    
    /**
     * Check if a ship can be placed horizontally starting at the given position
     */
    private fun canPlaceShipHorizontally(
        row: Int, 
        col: Int, 
        shipSize: Int, 
        playerGrid: Grid
    ): Boolean {
        if (col + shipSize > GameConstants.GRID_SIZE) return false
        
        for (i in 0 until shipSize) {
            val checkCol = col + i
            if (playerGrid.isMiss(row, checkCol)) {
                return false
            }
        }
        return true
    }
    
    /**
     * Check if a ship can be placed vertically starting at the given position
     */
    private fun canPlaceShipVertically(
        row: Int, 
        col: Int, 
        shipSize: Int, 
        playerGrid: Grid
    ): Boolean {
        if (row + shipSize > GameConstants.GRID_SIZE) return false
        
        for (i in 0 until shipSize) {
            val checkRow = row + i
            if (playerGrid.isMiss(checkRow, col)) {
                return false
            }
        }
        return true
    }
    
    /**
     * Get remaining ship sizes that haven't been sunk
     */
    private fun getRemainingShipSizes(aiFleet: Fleet): List<Int> {
        // In a real game, we'd track which ships have been sunk
        // For now, assume all ship sizes are still possible
        return ShipType.allShips().map { it.size }
    }
    
    /**
     * Apply hunt mode bonuses to probability grid
     */
    private fun applyHuntModeBonuses(playerGrid: Grid) {
        // Increase probability for cells adjacent to hits
        for (row in 0 until GameConstants.GRID_SIZE) {
            for (col in 0 until GameConstants.GRID_SIZE) {
                if (playerGrid.isHit(row, col) && !playerGrid.isSunk(row, col)) {
                    val adjacentPositions = GameConstants.getAdjacentPositions(row, col)
                    for ((adjRow, adjCol) in adjacentPositions) {
                        if (!playerGrid.hasBeenShot(adjRow, adjCol)) {
                            probabilityGrid[adjRow][adjCol] += 50
                        }
                    }
                }
            }
        }
    }
    
    /**
     * Apply strategic bonuses to probability grid
     */
    private fun applyStrategicBonuses(playerGrid: Grid) {
        // Center bias - cells in the center are more likely to be part of ships
        for (row in 0 until GameConstants.GRID_SIZE) {
            for (col in 0 until GameConstants.GRID_SIZE) {
                if (!playerGrid.hasBeenShot(row, col)) {
                    val centerDistance = kotlin.math.abs(row - 4.5) + kotlin.math.abs(col - 4.5)
                    val centerBonus = max(0, (10 - centerDistance).toInt())
                    probabilityGrid[row][col] += centerBonus
                }
            }
        }
        
        // Checkerboard pattern bonus for early game
        if (shotCount < 20) {
            for (row in 0 until GameConstants.GRID_SIZE) {
                for (col in 0 until GameConstants.GRID_SIZE) {
                    if (!playerGrid.hasBeenShot(row, col) && (row + col) % 2 == 0) {
                        probabilityGrid[row][col] += 5
                    }
                }
            }
        }
    }
    
    /**
     * Mark area around sunk ship as low priority
     */
    private fun markSunkShipArea(ship: Ship) {
        for ((row, col) in ship.getPositions()) {
            val adjacentPositions = GameConstants.getAdjacentPositions(row, col)
            for ((adjRow, adjCol) in adjacentPositions) {
                probabilityGrid[adjRow][adjCol] = max(0, probabilityGrid[adjRow][adjCol] - 25)
            }
        }
    }
    
    /**
     * Clear the probability grid
     */
    private fun clearProbabilityGrid() {
        for (row in 0 until GameConstants.GRID_SIZE) {
            for (col in 0 until GameConstants.GRID_SIZE) {
                probabilityGrid[row][col] = 0
            }
        }
    }
    
    /**
     * Get AI statistics
     */
    fun getStats(): AIStats {
        val accuracy = if (shotCount > 0) hitCount.toFloat() / shotCount else 0f
        return AIStats(
            shotCount = shotCount,
            hitCount = hitCount,
            accuracy = accuracy,
            difficulty = difficulty,
            huntMode = huntMode
        )
    }
    
    /**
     * Get current probability grid (for debugging/visualization)
     */
    fun getProbabilityGrid(): Array<Array<Int>> {
        return probabilityGrid.map { it.clone() }.toTypedArray()
    }
    
    /**
     * Auto-place ships for AI with strategic positioning
     */
    fun autoPlaceShips(fleet: Fleet) {
        fleet.clearPlacements()
        val ships = fleet.getAllShips().sortedByDescending { it.size }
        
        for (ship in ships) {
            var attempts = 0
            var placed = false
            
            while (!placed && attempts < 1000) {
                val row = Random.nextInt(GameConstants.GRID_SIZE)
                val col = Random.nextInt(GameConstants.GRID_SIZE)
                val orientation = if (Random.nextBoolean()) Orientation.HORIZONTAL else Orientation.VERTICAL
                
                if (fleet.canPlaceShipAt(ship, row, col, orientation)) {
                    fleet.placeShip(ship, row, col, orientation)
                    placed = true
                }
                attempts++
            }
        }
    }
}

/**
 * AI statistics
 */
data class AIStats(
    val shotCount: Int,
    val hitCount: Int,
    val accuracy: Float,
    val difficulty: AIDifficulty,
    val huntMode: Boolean
) 